dulwich/credentials.py

```diff
@@ -0,0 +1,295 @@
+
+        helper_input.append("")
+
+        try:
+            res = subprocess.run(  # type: ignore # breaks on 3.6
+                cmd,
+                check=True,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                input=os.linesep.join(helper_input).encode("ascii"),
+                **self._run_kwargs,
+            )
+        except subprocess.CalledProcessError as exc:
+            raise CredentialNotFoundError(exc.stderr) from exc
+        except FileNotFoundError as exc:
+            raise CredentialNotFoundError("Helper not found") from exc
+
+        credentials = {}
+        for line in res.stdout.strip().splitlines():
+            try:
+                key, value = line.split(b"=")
+                credentials[key] = value
+            except ValueError:
+                continue
+
+        if not all(
+            (credentials, b"username" in credentials, b"password" in credentials)
+        ):
+            raise CredentialNotFoundError("Could not get credentials from helper")
+
+        return credentials[b"username"], credentials[b"password"]
+
+    def store(self, *args, **kwargs):
+        """Store the credential, if applicable to the helper"""
+        raise NotImplementedError
+
+    def erase(self, *args, **kwargs):
+        """Remove a matching credential, if any, from the helper’s storage"""
+        raise NotImplementedError
+
+    @classmethod
+    def from_config(
+        cls, config: ConfigDict, url: Optional[str] = None
+    ) -> "CredentialHelper":
+        encoding = config.encoding or sys.getdefaultencoding()
+        # We will try to match credential section's url with the given url,
+        # falling back to the generic section if there's no match or
+        # if url is None
+        section = urlmatch_credential_section(config, url)
+
+        command = config.get(section, b"helper")
+        assert command and isinstance(command, bytes)
+        return cls(command.decode(encoding))
+
+
+def get_credentials_from_store(
+    scheme: bytes,
+    hostname: bytes,
+    username: Optional[bytes] = None,
+    fnames: List[str] = DEFAULT_GIT_CREDENTIALS_PATHS,
+):
+    for fname in fnames:
+        try:
+            with open(fname, "rb") as f:
+                for line in f:
+                    parsed_line = urlparse(line.strip())
+                    if (
+                        parsed_line.scheme == scheme
+                        and parsed_line.hostname == hostname
+                        and (username is None or parsed_line.username == username)
+                    ):
+                        return parsed_line.username, parsed_line.password
+        except FileNotFoundError:
+            # If the file doesn't exist, try the next one.
+            continue
+
+
+def get_credentials_from_helper(base_url: str, config) -> Tuple[bytes, bytes]:
+    """Retrieves credentials for the given url from git credential helpers"""
+    if isinstance(config, StackedConfig):
```

Comment:
Given an URL, git loops though the available configs and matches urls in credentials sections with the provided URL, which is something we cannot do with the config API.

For example with a config like this:
```
[credential "https://github.com/jelmer/dulwich"]
    helper = githubhelper

[credential]
    helper = generichelper
```
Running 
```python
config.get((b"credential", b"https://github.com))
```

would return `generichelper` instead of `githubhelper`, which is not what we want.

I refactored this a bit so that we use `urlmatch_credential_sections` to cycle though config sections that match the URL.
This could be extracted to become a config API method, although I'm not sure if there's any other use cases where one needs to match URLs in sections, apart from the credential helpers scenario.